### PR TITLE
add constraint to scipy to avoid broken openblas for osx-arm

### PR DESCRIPTION
**Pull request by h-vetinari**
Requested by the upstream maintainers [here](https://github.com/scipy/scipy/issues/16767#issuecomment-1205207606), CC @rgommers.

Openblas 0.3.20 has been [live](https://anaconda.org/conda-forge/openblas/files) for almost 4 months, but it only has ~9000 downloads on osx-arm. I'm not excited about breaking people's environment reproducibility, but silently wrong results seem worse.

<details>
<summary>Output of <code>python show_diff.py</code></summary>

```
>python show_diff.py
Downloading: https://conda.anaconda.org/conda-forge/noarch/repodata_from_packages.json.bz2
Downloading: https://conda.anaconda.org/conda-forge/noarch/repodata.json.bz2
Downloading: https://conda.anaconda.org/conda-forge/linux-64/repodata_from_packages.json.bz2
Downloading: https://conda.anaconda.org/conda-forge/linux-64/repodata.json.bz2
Downloading: https://conda.anaconda.org/conda-forge/linux-armv7l/repodata_from_packages.json.bz2
Downloading: https://conda.anaconda.org/conda-forge/linux-armv7l/repodata.json.bz2
Downloading: https://conda.anaconda.org/conda-forge/linux-aarch64/repodata_from_packages.json.bz2
Downloading: https://conda.anaconda.org/conda-forge/linux-aarch64/repodata.json.bz2
Downloading: https://conda.anaconda.org/conda-forge/linux-ppc64le/repodata_from_packages.json.bz2
Downloading: https://conda.anaconda.org/conda-forge/linux-ppc64le/repodata.json.bz2
Downloading: https://conda.anaconda.org/conda-forge/osx-64/repodata_from_packages.json.bz2
Downloading: https://conda.anaconda.org/conda-forge/osx-64/repodata.json.bz2
Downloading: https://conda.anaconda.org/conda-forge/osx-arm64/repodata_from_packages.json.bz2
Downloading: https://conda.anaconda.org/conda-forge/osx-arm64/repodata.json.bz2
osx-arm64::scipy-1.5.2-py38h2dbcbbe_2.tar.bz2
-  "version": "1.5.2"
+  "version": "1.5.2",
+  "constrains": [
+    "openblas <0.3.20"
+  ]
osx-arm64::scipy-1.5.2-py38hdf044fb_2.tar.bz2
-  "version": "1.5.2"
+  "version": "1.5.2",
+  "constrains": [
+    "openblas <0.3.20"
+  ]
osx-arm64::scipy-1.5.2-py39h39b03cc_2.tar.bz2
-  "version": "1.5.2"
+  "version": "1.5.2",
+  "constrains": [
+    "openblas <0.3.20"
+  ]
osx-arm64::scipy-1.5.2-py39h73ea49b_2.tar.bz2
-  "version": "1.5.2"
+  "version": "1.5.2",
+  "constrains": [
+    "openblas <0.3.20"
+  ]
osx-arm64::scipy-1.5.3-py38hd0c9ec0_1.tar.bz2
-  "version": "1.5.3"
+  "version": "1.5.3",
+  "constrains": [
+    "openblas <0.3.20"
+  ]
osx-arm64::scipy-1.5.3-py38hdf044fb_0.tar.bz2
-  "version": "1.5.3"
+  "version": "1.5.3",
+  "constrains": [
+    "openblas <0.3.20"
+  ]
osx-arm64::scipy-1.5.3-py39h5060c3b_1.tar.bz2
-  "version": "1.5.3"
+  "version": "1.5.3",
+  "constrains": [
+    "openblas <0.3.20"
+  ]
osx-arm64::scipy-1.5.3-py39h73ea49b_0.tar.bz2
-  "version": "1.5.3"
+  "version": "1.5.3",
+  "constrains": [
+    "openblas <0.3.20"
+  ]
osx-arm64::scipy-1.6.0-py38hdf044fb_0.tar.bz2
-  "version": "1.6.0"
+  "version": "1.6.0",
+  "constrains": [
+    "openblas <0.3.20"
+  ]
osx-arm64::scipy-1.6.0-py39h73ea49b_0.tar.bz2
-  "version": "1.6.0"
+  "version": "1.6.0",
+  "constrains": [
+    "openblas <0.3.20"
+  ]
osx-arm64::scipy-1.6.1-py38hdf044fb_0.tar.bz2
-  "version": "1.6.1"
+  "version": "1.6.1",
+  "constrains": [
+    "openblas <0.3.20"
+  ]
osx-arm64::scipy-1.6.1-py39h73ea49b_0.tar.bz2
-  "version": "1.6.1"
+  "version": "1.6.1",
+  "constrains": [
+    "openblas <0.3.20"
+  ]
osx-arm64::scipy-1.6.2-py38hd0c9ec0_0.tar.bz2
-  "version": "1.6.2"
+  "version": "1.6.2",
+  "constrains": [
+    "openblas <0.3.20"
+  ]
osx-arm64::scipy-1.6.2-py39h5060c3b_0.tar.bz2
-  "version": "1.6.2"
+  "version": "1.6.2",
+  "constrains": [
+    "openblas <0.3.20"
+  ]
osx-arm64::scipy-1.6.3-py38hd0c9ec0_0.tar.bz2
-  "version": "1.6.3"
+  "version": "1.6.3",
+  "constrains": [
+    "openblas <0.3.20"
+  ]
osx-arm64::scipy-1.6.3-py39h5060c3b_0.tar.bz2
-  "version": "1.6.3"
+  "version": "1.6.3",
+  "constrains": [
+    "openblas <0.3.20"
+  ]
osx-arm64::scipy-1.7.0-py38hd0c9ec0_0.tar.bz2
-  "version": "1.7.0"
+  "version": "1.7.0",
+  "constrains": [
+    "openblas <0.3.20"
+  ]
osx-arm64::scipy-1.7.0-py39h5060c3b_0.tar.bz2
-  "version": "1.7.0"
+  "version": "1.7.0",
+  "constrains": [
+    "openblas <0.3.20"
+  ]
osx-arm64::scipy-1.7.1-py38hd0c9ec0_0.tar.bz2
-  "version": "1.7.1"
+  "version": "1.7.1",
+  "constrains": [
+    "openblas <0.3.20"
+  ]
osx-arm64::scipy-1.7.1-py39h5060c3b_0.tar.bz2
-  "version": "1.7.1"
+  "version": "1.7.1",
+  "constrains": [
+    "openblas <0.3.20"
+  ]
osx-arm64::scipy-1.7.2-py310h6ecf4ae_0.tar.bz2
-  "version": "1.7.2"
+  "version": "1.7.2",
+  "constrains": [
+    "openblas <0.3.20"
+  ]
osx-arm64::scipy-1.7.2-py38hd0c9ec0_0.tar.bz2
-  "version": "1.7.2"
+  "version": "1.7.2",
+  "constrains": [
+    "openblas <0.3.20"
+  ]
osx-arm64::scipy-1.7.2-py39h5060c3b_0.tar.bz2
-  "version": "1.7.2"
+  "version": "1.7.2",
+  "constrains": [
+    "openblas <0.3.20"
+  ]
osx-arm64::scipy-1.7.3-py310h6ecf4ae_0.tar.bz2
-  "version": "1.7.3"
+  "version": "1.7.3",
+  "constrains": [
+    "openblas <0.3.20"
+  ]
osx-arm64::scipy-1.7.3-py38hd0c9ec0_0.tar.bz2
-  "version": "1.7.3"
+  "version": "1.7.3",
+  "constrains": [
+    "openblas <0.3.20"
+  ]
osx-arm64::scipy-1.7.3-py39h5060c3b_0.tar.bz2
-  "version": "1.7.3"
+  "version": "1.7.3",
+  "constrains": [
+    "openblas <0.3.20"
+  ]
osx-arm64::scipy-1.8.0-py310h6ecf4ae_0.tar.bz2
-  "version": "1.8.0"
+  "version": "1.8.0",
+  "constrains": [
+    "openblas <0.3.20"
+  ]
osx-arm64::scipy-1.8.0-py310h6ecf4ae_1.tar.bz2
-  "version": "1.8.0"
+  "version": "1.8.0",
+  "constrains": [
+    "openblas <0.3.20"
+  ]
osx-arm64::scipy-1.8.0-py38hd0c9ec0_0.tar.bz2
-  "version": "1.8.0"
+  "version": "1.8.0",
+  "constrains": [
+    "openblas <0.3.20"
+  ]
osx-arm64::scipy-1.8.0-py38hd0c9ec0_1.tar.bz2
-  "version": "1.8.0"
+  "version": "1.8.0",
+  "constrains": [
+    "openblas <0.3.20"
+  ]
osx-arm64::scipy-1.8.0-py39h5060c3b_0.tar.bz2
-  "version": "1.8.0"
+  "version": "1.8.0",
+  "constrains": [
+    "openblas <0.3.20"
+  ]
osx-arm64::scipy-1.8.0-py39h5060c3b_1.tar.bz2
-  "version": "1.8.0"
+  "version": "1.8.0",
+  "constrains": [
+    "openblas <0.3.20"
+  ]
osx-arm64::scipy-1.8.1-py310hdb41229_0.tar.bz2
-  "version": "1.8.1"
+  "version": "1.8.1",
+  "constrains": [
+    "openblas <0.3.20"
+  ]
osx-arm64::scipy-1.8.1-py310hdb41229_2.tar.bz2
-  "version": "1.8.1"
+  "version": "1.8.1",
+  "constrains": [
+    "openblas <0.3.20"
+  ]
osx-arm64::scipy-1.8.1-py38h4f188a7_0.tar.bz2
-  "version": "1.8.1"
+  "version": "1.8.1",
+  "constrains": [
+    "openblas <0.3.20"
+  ]
osx-arm64::scipy-1.8.1-py38h4f188a7_2.tar.bz2
-  "version": "1.8.1"
+  "version": "1.8.1",
+  "constrains": [
+    "openblas <0.3.20"
+  ]
osx-arm64::scipy-1.8.1-py39h14896cb_0.tar.bz2
-  "version": "1.8.1"
+  "version": "1.8.1",
+  "constrains": [
+    "openblas <0.3.20"
+  ]
osx-arm64::scipy-1.8.1-py39h14896cb_2.tar.bz2
-  "version": "1.8.1"
+  "version": "1.8.1",
+  "constrains": [
+    "openblas <0.3.20"
+  ]
osx-arm64::scipy-1.9.0-py310hdb41229_0.tar.bz2
-  "version": "1.9.0"
+  "version": "1.9.0",
+  "constrains": [
+    "openblas <0.3.20"
+  ]
osx-arm64::scipy-1.9.0-py38h4f188a7_0.tar.bz2
-  "version": "1.9.0"
+  "version": "1.9.0",
+  "constrains": [
+    "openblas <0.3.20"
+  ]
osx-arm64::scipy-1.9.0-py39h14896cb_0.tar.bz2
-  "version": "1.9.0"
+  "version": "1.9.0",
+  "constrains": [
+    "openblas <0.3.20"
+  ]
Downloading: https://conda.anaconda.org/conda-forge/win-32/repodata_from_packages.json.bz2
Downloading: https://conda.anaconda.org/conda-forge/win-32/repodata.json.bz2
Downloading: https://conda.anaconda.org/conda-forge/win-64/repodata_from_packages.json.bz2
Downloading: https://conda.anaconda.org/conda-forge/win-64/repodata.json.bz2
```

</details>

### CODE REVIEW

**Comment by rgommers** [comment]
@h-vetinari I think `0.3.19` has the same problem: https://github.com/numpy/numpy/issues/21950#issuecomment-1180435643

**Comment by rgommers** [comment]
That is also a more convincing demonstration of a problem - the results of a `numpy.linalg.eigh` are clearly incorrect, while with `0.3.18` from conda-forge they work fine.

**Comment by h-vetinari** [comment]
> @h-vetinari I think `0.3.19` has the same problem: [numpy/numpy#21950 (comment)](https://github.com/numpy/numpy/issues/21950#issuecomment-1180435643)

Fair point, though since we don't have builds of 0.3.19 (and aren't going to add them with near certainty) this is functionally equivalent. Happy to adapt though if someone feels strongly about it.

**Comment by h-vetinari** [comment]
Gentle ping @conda-forge/core

**Comment by isuruf** [comment]
The issue in openblas for m1 was patched in openblas 0.3.20 build 1 in conda-forge. So @rgommers analysis is incorrect and should be reported upstream.

**Comment by h-vetinari** [comment]
Thanks for chiming in. It seems there are _additional_ problems with 0.3.20 on osx-arm that don't appear with 0.3.18, so more upstream reporting notwithstanding, I think this PR solves real issues being encountered at the moment.

**Comment by rgommers** [comment]
> The issue in openblas for m1 was patched in openblas 0.3.20 build 1 in conda-forge. So @rgommers analysis is incorrect and should be reported upstream.

I just tried again, the reproducer I linked to is still broken with build 1:
```
% mamba list openblas
# packages in environment at /Users/rgommers/mambaforge/envs/scipy-dev:
#
# Name                    Version                   Build  Channel
libopenblas               0.3.20          openmp_h130de29_1    conda-forge
openblas                  0.3.20          openmp_h7efc2c9_1    conda-forge

% python m1_openblas_reproducer.py
-1.216802553763662
(-0.9999999999999998-2.2371143170757382e-17j)
(-1.707106781186547-3.251767952832691e-17j)
```
with `0.3.18` things look fine.

**Comment by h-vetinari** [comment]
openblas 0.3.21 is now live - which should alleviate this somewhat.

**Comment by h-vetinari** [comment]
I just opened https://github.com/conda-forge/openblas-feedstock/issues/143 to start collecting some of the known correctness issues with 0.3.20 & 0.3.21. A lot of these have follow-on issues in different trackers because openblas/numpy/scipy are so low in the stack.

I think we shouldn't be pushing subtly broken builds, and that we really should pin to a known-working solution while these things are worked out. PTAL @conda-forge/core

File: recipe/gen_patch_json.py [hunk 1]
```diff
@@ -738,6 +738,14 @@ def _gen_new_index_per_key(repodata, subdir, index_key):
             depends.append("_r-mutex 1.* anacondar_1")
             record["depends"] = depends
 
+        if (record_name == "scipy" and subdir == "osx-arm64"
+                and record.get('timestamp', 0) < 1659636598216):
```
Comment (h-vetinari) [inline recipe/gen_patch_json.py:742]:
```suggestion
                and record.get('timestamp', 0) < 1660645059386):
```

File: recipe/gen_patch_json.py [hunk 2]
```diff
@@ -738,6 +738,14 @@ def _gen_new_index_per_key(repodata, subdir, index_key):
...
+            # openblas 0.3.20 is broken with osx-arm, see
+            # https://github.com/scipy/scipy/issues/16767
+            new_constrains = record.get('constrains', [])
+            new_constrains.append("openblas <0.3.20")
+            record['constrains'] = new_constrains
```
Comment (h-vetinari) [inline recipe/gen_patch_json.py:747]:
```suggestion
            # openblas 0.3.20 & 0.3.21 are broken with osx-arm, see
            # https://github.com/conda-forge/openblas-feedstock/issues/143
            new_constrains = record.get('constrains', [])
            new_constrains.append("openblas <0.3.19")
            record['constrains'] = new_constrains
```

**Comment by h-vetinari** [comment]
Closing as obsolete

**Pull request closed without merging**